src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                }
+                else
+                {
+                    _lineBytePosition++;
+                }
+            }
+        }
+
+        /// <summary>
+        /// This method contains the logic for processing the next value token and determining
+        /// what type of data it is.
+        /// </summary>
+        private bool ConsumeValue(byte marker)
+        {
+            if (marker == JsonConstants.Quote)
+                return ConsumeString();
+            else if (marker == JsonConstants.OpenBrace)
+                StartObject();
+            else if (marker == JsonConstants.OpenBracket)
+                StartArray();
+            else if ((uint)(marker - '0') <= '9' - '0' || marker == '-')
+                return ConsumeNumber();
+            else if (marker == 'f')
+                return ConsumeLiteral(JsonConstants.FalseValue, JsonTokenType.False);
+            else if (marker == 't')
+                return ConsumeLiteral(JsonConstants.TrueValue, JsonTokenType.True);
+            else if (marker == 'n')
+                return ConsumeLiteral(JsonConstants.NullValue, JsonTokenType.Null);
+            else
+            {
+                if (_readerOptions.CommentHandling != JsonCommentHandling.Default)
+                {
+                    if (_readerOptions.CommentHandling == JsonCommentHandling.AllowComments)
+                    {
+                        if (marker == JsonConstants.Solidus)
+                        {
+                            return ConsumeComment();
+                        }
+                    }
+                    else
+                    {
+                        // JsonCommentHandling.SkipComments
+                        if (marker == JsonConstants.Solidus)
+                        {
+                            if (SkipComment())
+                            {
+                                if (_consumed >= (uint)_buffer.Length)
+                                {
+                                    if (_isNotPrimitive && IsLastSpan)
+                                    {
+                                        if (_tokenType != JsonTokenType.EndArray && _tokenType != JsonTokenType.EndObject)
+                                            ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidEndOfJson);
+                                    }
+                                    return false;
+                                }
+
+                                byte first = _buffer[_consumed];
+
+                                if (first <= JsonConstants.Space)
+                                {
+                                    SkipWhiteSpace();
+                                    if (!HasMoreData())
+                                        return false;
+                                    first = _buffer[_consumed];
+                                }
+
+                                return ConsumeValue(first);
```

Comment:
This seems like it could be iterative (even with "marker = first; goto Begin;") instead of recursive.